param.go

```diff
@@ -0,0 +1,187 @@
+package dig
+
+import (
+	"fmt"
+	"reflect"
+)
+
+// The param interface represents a dependency for a constructor.
+//
+// The following implementations exist:
+//  paramList     All arguments of the constructor.
+//  paramSimple   An explicitly requested type.
+//  paramObject   dig.In struct where each field in the struct can be another
+//                param.
+type (
+	param interface {
+		param()
+
+		// Human-friendly name for this param type. We use this exclusively in
+		// error messages.
+		ParamTypeName() string
+
+		// Comprehensive list of dependencies this parameter represents.
+		Dependencies() []edge
+	}
+
+	// paramList holds all arguments of the constructor as params.
+	paramList struct {
+		ctype reflect.Type // type of the constructor
+
+		Params []param
+	}
+
+	// paramSimple is an explicitly requested type, optionally with a name.
+	//
+	// This object must be present in the graph as-is unless it's specified as
+	// optional.
+	paramSimple struct {
+		Name     string
+		Optional bool
+		Type     reflect.Type
+	}
+
+	// paramObject is a dig.In struct where each field is another param.
+	//
+	// This object is not expected in the graph as-is.
+	paramObject struct {
+		Type   reflect.Type
+		Fields []paramObjectField
+
+		deps []edge
+	}
+
+	// paramObjectField is a single field of a dig.In struct.
+	paramObjectField struct {
+		// Name of the field specified via a name:".." tag.
+		Name string
```

Comment:
We already cover this in the paramSimple field, do we need it here?

Or, if we want to go the other way, should we pull the Optional field to this level too, and remove the Name/Optional fields from the paramSimple?

Reply:
Documentation here was wrong. This is the name of the field in the struct. Fixed. Thanks for catching.